datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -2645,6 +2643,106 @@ pub struct Union {
     pub schema: DFSchemaRef,
 }
 
+impl Union {
+    /// Constructs new Union instance deriving schema from inputs.
+    fn try_new(inputs: Vec<Arc<LogicalPlan>>) -> Result<Self> {
+        let schema = Self::derive_schema_from_inputs(&inputs, false)?;
+        Ok(Union { inputs, schema })
+    }
+
+    /// Constructs new Union instance deriving schema from inputs.
+    /// Inputs do not have to have matching types and produced schema will
+    /// take type from the first input.
+    pub fn try_new_with_loose_types(inputs: Vec<Arc<LogicalPlan>>) -> Result<Self> {
```

Comment:
Another name for this one is  `try_new_with_coerce_types` to emphasize it is coercing the types

Reply:
I believe this one should be coercing types, but it's not doing this. It just takes a type from first UNION branch.
This is not new logic though. It's just moved from plan builder over here, to avoid duplicating code.

I didn't know why this logic existed in this shape, but it felt intentional enough not to simply replace it **in this PR**. I would prefer it to be fixed later... Hence this function name to make the caller wonder what "loose types" mean.

Reply:
ok

Reply:
Yeah, [this was noted once before](https://github.com/apache/datafusion/pull/12142#discussion_r1730651101). Thank you for making it more explicit.

Reply:
-> https://github.com/apache/datafusion/issues/14380